Added KAG server list

eps0003 opened the pull request:
Although this repo doesn't include KAG-related aspects of the bot, I hope you will consider manually adding this to the KAG version of the bot.

I added a channel which has a message with a list of populated KAG servers. Official servers are highlighted in red and other servers in yellow. The channel name is changed to state the number of servers and players that are active. The interval at which this updates can be changed in the config, as well as the channel and message IDs.

I had to add the !ping command to be able to initially create the server list message since Discord bots can only edit their own messages. It is only accessible to mods and responds with 'Pong!'.

I also made the bot's presence message more accurate by using data from the KAG API rather than the Steam statistics currently being used.

This would make a great addition to the server since it would allow people to easily see who is online without opening the game. It has been used in the Oceania KAG Discord server and people really like it.

eps0003 committed:
Added KAG server list and !ping command

eps0003 committed:
Pluralized presence message

eps0003 committed:
Fixed outdated presence method

eps0003 committed:
Fixed irrelevant grammar

Mazey commented:
PR is incompatible with bot rewrite. Requesting changes.

eps0003 committed:
Merge pull request #1 from Mazey/master

REWRITE

eps0003 committed:
Merge branch 'master' into server-list

eps0003 committed:
Recoded server list

eps0003 committed:
Removed token from config :sweat:

eps0003 committed:
Minor fixes

Mazey commented on modules/serverlist.js:
```diff
@@ -0,0 +1,81 @@
+		message = msg;
+		loop();
+	}).catch(err => {
+		return; // Don't do anything if the message doesn't exist
```
should create a message in the channel

eps0003 replied:
which channel?

Mazey replied:
the channel specified in the configs, in which you are checking for the message

eps0003 replied:
i can make it display a console log or even post a message in some sort of bot command log channel. the plan is to have the server list in its own channel with no other messages

Mazey replied:
I mean if I want to set this up, the easiest would be to create a new channel, set the channel id and start the bot. Now I have to start the bot, make the bot write a message in said channel, close the bot, update config, start the bot.

Mazey commented on modules/serverlist.js:
```diff
@@ -0,0 +1,81 @@
+
+			// Update message, channel and presence
+			channel.setName(`${servers.length}-${plural(servers.length, 'server')}_${players}-${plural(players, 'player')}`);
+			client.user.setPresence({ status: 'online', game: { name: `with ${players} ${plural(players, 'fishy', 'ies', 1)}` } });
```
so this replaces the playercount module? please remove it and exclude it from the config and index.js.

eps0003 replied:
would making it an option in config be sufficient? it uses the server data which would be hard to access from a different module.

eps0003 replied:
it is meant to completely replace the playercount module since thats quite inaccurate tbh

Mazey commented on modules/serverlist.js:
```diff
@@ -0,0 +1,81 @@
+// Made by epsilon#3892
+// https://forum.thd.vg/members/16800/
+
+var config = require("../config.json");
```
should be const

Mazey commented on modules/serverlist.js:
```diff
@@ -0,0 +1,81 @@
+// https://forum.thd.vg/members/16800/
+
+var config = require("../config.json");
+var request = require("request")
```
should be const

Mazey commented on modules/serverlist.js:
```diff
@@ -0,0 +1,81 @@
+		return; // Don't do anything if the message doesn't exist
+	});
+
+	function loop() {
```
change the function name to something less vague. Maybe refresh or refreshServers.

Mazey commented on modules/serverlist.js:
```diff
@@ -0,0 +1,81 @@
+var request = require("request")
+
+module.exports = function(client) {
+	var module = {};
```
you're not using this variable

eps0003 replied:
im using it to fetch the message (line 11)

Mazey replied:
I'm talking about the module variable

eps0003 replied:
don't i need that for when it is returned at the end? ive only just learnt how to use modules through this bot

Mazey commented on modules/serverlist.js:
```diff
@@ -0,0 +1,81 @@
+		request({
+		    url: 'https://api.kag2d.com/v1/game/thd/kag/servers?filters=[{"field":"current","op":"eq","value":"true"},{"field":"connectable","op":"eq","value":true},{"field":"currentPlayers","op":"gt","value":"0"}]',
+		    json: true
+		}, function (error, response, body) {
```
change this into an arrow function for consistency

eps0003 replied:
im using it to fetch the message (line 11)

Mazey commented on modules/serverlist.js:
```diff
@@ -0,0 +1,81 @@
+				let prefix = (/(?=^KAG Official \w+ (AUS|AU|EU|US|USA)\b)|(?=^Official Modded Server (AUS|AU|EU|US|USA)\b)/g.test(server.name)) ? '-' : '+';
+				let full = (server.playerPercentage >= 1) ? ' [FULL]' : '';
+				let specs = (server.spectatorPlayers > 0) ? ` (${server.spectatorPlayers} spec)` : '';
+				return `${prefix} ${alignText(server.name, 50, -1)} ${alignText(server.currentPlayers, 3, 1)}/${server.maxPlayers}${full}${specs}\n​${server.playerList.join('  ')}`;
```
use plural function  on spec here

eps0003 replied:
it was intentionally like that but sure

Mazey replied:
1 spec 2 spec 3 spec?

eps0003 replied:
'specs' looks like 'specs of dust', whereas 'spec' somewhat retains its meaning

Mazey commented on modules/serverlist.js:
```````diff
@@ -0,0 +1,81 @@
+			let players = servers.reduce((t, x) => t + x.currentPlayers, 0);
+			let text = '```md\n' + `# Server list - ${servers.length} ${plural(servers.length, 'server')}, ${players} ${plural(players, 'player')}` + '``````diff\n';
+			text += servers.map(server => {
+				let prefix = (/(?=^KAG Official \w+ (AUS|AU|EU|US|USA)\b)|(?=^Official Modded Server (AUS|AU|EU|US|USA)\b)/g.test(server.name)) ? '-' : '+';
```````
shouldn't you check official by IP?

eps0003 replied:
it should but that requires it to be updated for every new official server added. regex hasn't failed me yet, and if it does, ill just make it more strict

Mazey commented on modules/serverlist.js:
```diff
@@ -0,0 +1,81 @@
+		// Loop on an interval
+		let interval = config.server_list.interval * 1000;
+		let delay = interval - new Date() % interval;
+		setTimeout(module.loop, delay);
```
this part could be moved to an interval and loop is not a member of module it is simply `loop();`
and what is the point of the delay variable. just use interval?

eps0003 replied:
this interval method ensures it runs precisely every 10 seconds on the dot. this currently isn't necessary but is better than using setInterval

eps0003 committed:
Added !ping command, fixed suggestions

eps0003 committed:
Removed ping command and playercount module. Changed serverlist module

eps0003 committed:
Removed comments from testing

eps0003 committed:
Minor changes

- Replaced single quotes with double quotes for consistency
- Rewrote commands section of index.js (now allows prefixes that include spaces)
- Added a few more comments in serverlist module
- Removed steam_api_key property of config

eps0003 committed:
Removed module variable

eps0003 committed:
Removed variable for serverlist module
